sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamExecution.scala

```
@@ -0,0 +1,146 @@
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.execution.streaming
+
+import org.apache.spark.Logging
+import org.apache.spark.sql.catalyst.analysis.EliminateSubQueries
+import org.apache.spark.sql.{DataFrame, SQLContext}
+import org.apache.spark.sql.catalyst.plans.logical.{LocalRelation, LogicalPlan}
+import org.apache.spark.sql.execution.{SparkPlan, QueryExecution, LogicalRDD}
+
+/**
+ * Manages the execution of a streaming Spark SQL query that is occuring in a separate thread.
+ * Unlike a standard query, a streaming query executes repeatedly each time new data arrives at any
+ * [[Source]] present in the query plan. Whenever new data arrives, a [[QueryExecution]] is created
+ * and the results are committed transactionally to the given [[Sink]].
+ */
+class StreamExecution(
+    sqlContext: SQLContext,
+    private[sql] val logicalPlan: LogicalPlan,
+    val sink: Sink) extends Logging {
+
+  /** Tracks how much data we have processed from each input source. */
+  private[sql] val currentOffsets = new StreamProgress
+
+  /** All stream sources present the query plan. */
+  private val sources =
+    logicalPlan.collect { case s: StreamingRelation => s.source }
+
+  // Start the execution at the current Offset for the sink. (i.e. avoid reprocessing data
+  // that we have already processed).
+  {
+    val storedProgress = sink.currentProgress
+    sources.foreach { s =>
+      storedProgress.get(s).foreach { offset =>
+        currentOffsets.update(s, offset)
+      }
+    }
+  }
+
+  logInfo(s"Stream running at $currentOffsets")
+
+  /** When false, signals to the microBatchThread that it should stop running. */
+  @volatile private var shouldRun = true
+
+  /** The thread that runs the micro-batches of this stream. */
+  private[sql] val microBatchThread = new Thread("stream execution thread") {
+    override def run(): Unit = {
+      SQLContext.setActive(sqlContext)
+      while (shouldRun) { attemptBatch() }
```

Needs some exception handling. If this thread silently fails on some exception, what happens to execution? Is it terminated? How will the user get the exception. 

In SS, the solution was to forward any terminating exception to the thread stuck in streamingContext.awaitTermination().